tmt/utils.py

```diff
@@ -2475,8 +2475,12 @@ def _indent(iterable: Iterable[str]) -> Iterator[str]:
             else:
                 for line in item.splitlines():
                     yield f'{INDENT * " "}{line}'
-
-    yield click.style(str(exception), fg='red')
+    # XXX: We should handle option `--no-color`
+    no_color = True # FIXME: should get the no_color from tmt/__main__.py
```

Comment:
Can you try the following? `apply_colors_logging` should tell you whether to apply colors or not. It uses params from the current click context, and uses the right function to evaluate all inputs.

```python
click_context = click.get_current_context(silent=True)
_, apply_colors_logging = tmt.log.decide_colorization(click_context.params['no_color'], click_context.params['force_color'])
```

Reply:
Thanks very much for your quick feedback, @happz !

I had a try to get `click_context` in [render_exception()][01], but it is `None`. e.g.

```python
> /home/huanli/dev/2759/dev/tmt/utils.py(2480)render_exception()
-> click_context = click.get_current_context(silent=True)
(Pdb) l
2475 	            else:
2476 	                for line in item.splitlines():
2477 	                    yield f'{INDENT * " "}{line}'
2478 	    # XXX: We should handle option `--no-color`
2479 	    breakpoint()
2480 ->	    click_context = click.get_current_context(silent=True)
2481 	    _, apply_colors_logging = tmt.log.decide_colorization(
2482 	            click_context.params['no_color'],
2483 	            click_context.params['force_color'])
2484 	    #no_color = True  # FIXME: should get the no_color from tmt/__main__.py
2485 	    if not apply_colors_logging:
(Pdb) n
> /home/huanli/dev/2759/dev/tmt/utils.py(2481)render_exception()
-> _, apply_colors_logging = tmt.log.decide_colorization(
(Pdb) pp click_context
None
```

Looks we have to get `click_context` from [run_cli()][02]. What do you think?

[01]: https://github.com/teemtee/tmt/blob/main/tmt/utils.py#L2467
[02]: https://github.com/teemtee/tmt/blob/main/tmt/__main__.py#L4

Reply:
How did you run this code? Is it from a Python session, or from a genuine tmt command-line?

Reply:
Hi @happz, I ran the code via a genuine tmt command-line and the related code having `breakpoint()`. e.g.

```python
(dev) huanli@KVM:dev$ git diff
diff --git a/tmt/utils.py b/tmt/utils.py
index 047713c5..0c3c114a 100644
--- a/tmt/utils.py
+++ b/tmt/utils.py
@@ -2476,8 +2476,13 @@ def render_exception(exception: BaseException) -> Iterator[str]:
                 for line in item.splitlines():
                     yield f'{INDENT * " "}{line}'
     # XXX: We should handle option `--no-color`
-    no_color = True # FIXME: should get the no_color from tmt/__main__.py
-    if no_color:
+    breakpoint()
+    click_context = click.get_current_context(silent=True)
+    _, apply_colors_logging = tmt.log.decide_colorization(
+            click_context.params['no_color'],
+            click_context.params['force_color'])
+    #no_color = True  # FIXME: should get the no_color from tmt/__main__.py
+    if not apply_colors_logging:
         yield str(exception)
     else:
         yield click.style(str(exception), fg='red')

(dev) huanli@KVM:dev$ tmt --no-color run discover -h fmf -u https://127.0.0.1/nonexistent
/var/tmp/tmt/run-001

/plans/features/advanced
    discover
        how: fmf
        url: https://127.0.0.1/nonexistent
        fail: Command 'git clone https://127.0.0.1/nonexistent /var/tmp/tmt/run-001/plans/features/advanced/discover/default-0/tests' returned 128.
        fail: Command 'git clone https://127.0.0.1/nonexistent /var/tmp/tmt/run-001/plans/features/advanced/discover/default-0/tests' returned 128.
        fail: Command 'git clone https://127.0.0.1/nonexistent /var/tmp/tmt/run-001/plans/features/advanced/discover/default-0/tests' returned 128.

> /home/huanli/dev/2759/dev/tmt/utils.py(2480)render_exception()
-> click_context = click.get_current_context(silent=True)
(Pdb) l
2475 	            else:
2476 	                for line in item.splitlines():
2477 	                    yield f'{INDENT * " "}{line}'
2478 	    # XXX: We should handle option `--no-color`
2479 	    breakpoint()
2480 ->	    click_context = click.get_current_context(silent=True)
2481 	    _, apply_colors_logging = tmt.log.decide_colorization(
2482 	            click_context.params['no_color'],
2483 	            click_context.params['force_color'])
2484 	    #no_color = True  # FIXME: should get the no_color from tmt/__main__.py
2485 	    if not apply_colors_logging:
(Pdb) n
> /home/huanli/dev/2759/dev/tmt/utils.py(2481)render_exception()
-> _, apply_colors_logging = tmt.log.decide_colorization(
(Pdb) p click_context
None
######## <<<<<<<<<<<<<<<<<<  !!! click_context is None !!!


(Pdb) n
> /home/huanli/dev/2759/dev/tmt/utils.py(2482)render_exception()
-> click_context.params['no_color'],
(Pdb)
AttributeError: 'NoneType' object has no attribute 'params'
> /home/huanli/dev/2759/dev/tmt/utils.py(2482)render_exception()
-> click_context.params['no_color'],
(Pdb)
```

Reply:
Aha, got it. At this point, we left the Click code, which explains why Click cleaned up the context stack, and there is no active context :/ Now it makes sense.

Let me try a few experiments.

Reply:
Thanks very much for your solution, which is much better than mine. @happz